mindmeld/models/nn_utils/layers.py

```diff
@@ -0,0 +1,259 @@
+
+    nn_module = _get_module_or_attr("torch.nn", "Module")
+except ImportError:
+    nn_module = object
+    pass
+
+SEED = 7246
+
+logger = logging.getLogger(__name__)
+
+
+class EmbeddingLayer(nn_module):
+    """A pytorch wrapper layer for embeddings that takes input as a batched sequence of ids
+    and outputs embeddings correponding to those ids
+    """
+
+    def __init__(
+        self, num_tokens, emb_dim, padding_idx=None,
+        embedding_weights=None, update_embeddings=True,
+        embeddings_dropout=0.5, coefficients=None, update_coefficients=True
+    ):
+        super().__init__()
+
+        self.embeddings = nn.Embedding(num_tokens, emb_dim, padding_idx=padding_idx)
+        if embedding_weights is not None:
+            if isinstance(embedding_weights, dict):
+                # when weights are passed as dict with keys as indices and values as embeddings
+                for idx, emb in embedding_weights.items():
+                    self.embeddings.weight.data[idx] = torch.as_tensor(emb)
+                msg = f"Initialized {len(embedding_weights)} number of embedding weights " \
+                      f"from the embedder model"
+                logger.info(msg)
+            else:
+                # when weights are passed as an array or tensor
+                self.embeddings.load_state_dict({'weight': torch.as_tensor(embedding_weights)})
+        self.embeddings.weight.requires_grad = update_embeddings
+
+        self.embedding_for_coefficients = None
+        if coefficients is not None:
+            if not len(coefficients) == num_tokens:
+                msg = f"Length of coefficients ({len(coefficients)}) must match the number of " \
+                      f"embeddings ({num_tokens})"
+                raise ValueError(msg)
+            self.embedding_for_coefficients = nn.Embedding(num_tokens, 1, padding_idx=padding_idx)
+            self.embedding_for_coefficients.load_state_dict({'weight': coefficients})
+            self.embedding_for_coefficients.weight.requires_grad = update_coefficients
+
+        self.dropout = nn.Dropout(embeddings_dropout)
+
+    def forward(self, padded_token_ids):
+        # padded_token_ids: dim: [BS, SEQ_LEN]
+        # returns:          dim: [BS, SEQ_LEN, EMB_DIM]
+
+        # [BS, SEQ_LEN] -> [BS, SEQ_LEN, EMB_DIM]
+        outputs = self.embeddings(padded_token_ids)
+
+        if self.embedding_for_coefficients:
+            # [BS, SEQ_LEN] -> [BS, SEQ_LEN, 1]
+            coefficients = self.embedding_for_coefficients(padded_token_ids)
+            # [BS, SEQ_LEN, EMB_DIM] -> [BS, SEQ_LEN, EMB_DIM]
+            outputs = torch.mul(outputs, coefficients)
+
+        outputs = self.dropout(outputs)
+
+        return outputs
+
+
+class CnnLayer(nn_module):
+
+    def __init__(self, emb_dim, kernel_sizes, num_kernels):
+        super().__init__()
+
+        if isinstance(num_kernels, list) and len(num_kernels) != len(kernel_sizes):
```

Comment:
Can you please clarify why we are taking the `len(num_kernels)`? Since 'num_var' normally refers to an int. Would be helpful here to add a comment with the type of `num_kernels` or rename it

Reply:
Sure, I'll see how I can change. I will provide type hints that num_kernels is a List specifying number of kernels per kernel size specified.